src/server/pfs/server/master.go

```diff
@@ -69,42 +87,67 @@ func (d *driver) master(ctx context.Context) {
 	})
 }
 
-func (d *driver) finishCommits(ctx context.Context) error {
+func (d *driver) finishCommits(ctx context.Context) (retErr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 	repos := make(map[string]context.CancelFunc)
 	defer func() {
 		for _, cancel := range repos {
 			cancel()
 		}
 	}()
 	compactor := newCompactor(d.storage, d.env.StorageConfig.StorageCompactionMaxFanIn)
-	err := d.repos.ReadOnly(ctx).WatchF(func(ev *watch.Event) error {
-		if ev.Type == watch.EventError {
-			return ev.Err
-		}
-		key := string(ev.Key)
-		if ev.Type == watch.EventDelete {
-			if cancel, ok := repos[key]; ok {
-				cancel()
-			}
-			delete(repos, key)
-			return nil
-		}
-		if _, ok := repos[key]; ok {
-			return nil
-		}
-		ctx, cancel := context.WithCancel(ctx)
-		repos[key] = cancel
-		go func() {
-			backoff.RetryUntilCancel(ctx, func() error { //nolint:errcheck
-				return d.finishRepoCommits(ctx, compactor, key)
-			}, backoff.NewInfiniteBackOff(), func(err error, d time.Duration) error {
-				log.Errorf("error finishing commits for repo %v: %v, retrying in %v", key, err, d)
+	return consistenthashing.WithRing(ctx, d.etcdClient, d.log, path.Join(d.prefix, masterLockPath, "ring"),
+		func(ctx context.Context, ring *consistenthashing.Ring) error {
+			err := d.repos.ReadOnly(ctx).WatchF(func(ev *watch.Event) error {
+				if ev.Type == watch.EventError {
+					return ev.Err
+				}
+				key := string(ev.Key)
+				lockPrefix := path.Join("repos", key)
+				if ev.Type == watch.EventDelete {
+					if cancel, ok := repos[key]; ok {
+						if err := ring.Unlock(lockPrefix); err != nil {
+							return err
+						}
+						cancel()
+					}
+					delete(repos, key)
+					return nil
+				}
+				if _, ok := repos[key]; ok {
+					return nil
+				}
+				ctx, cancel := context.WithCancel(ctx)
+				repos[key] = cancel
+				go func() {
+					backoff.RetryUntilCancel(ctx, func() error { //nolint:errcheck
+						ctx, cancel := context.WithCancel(ctx)
+						defer cancel()
+						lockCtx, err := ring.Lock(ctx, lockPrefix)
+						if err != nil {
+							return errors.Wrap(err, "error locking repo lock")
+						}
+						defer func() {
+							if err := ring.Unlock(lockPrefix); err != nil {
+								if retErr == nil {
```

Comment:
I just realized this can be simplified further to just `retErr = multierror.Append...`. It is fine if `retErr` is nil because append will only capture the unlock error in that case.